core/store/src/db.rs

```diff
@@ -0,0 +1,187 @@
+use parking_lot::RwLock;
+use rocksdb::{
+    ColumnFamily, ColumnFamilyDescriptor, DBCompactionStyle, DBIterator, IteratorMode, Options,
+    ReadOptions, WriteBatch, DB,
+};
+use std::collections::HashMap;
+use std::io;
+
+fn other_io_err<E>(e: E) -> io::Error
+where
+    E: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    io::Error::new(io::ErrorKind::Other, e)
+}
+
+pub struct DBTransaction {
+    pub ops: Vec<DBOp>,
+}
+
+pub enum DBOp {
+    Insert { col: usize, key: Vec<u8>, value: Vec<u8> },
+    Delete { col: usize, key: Vec<u8> },
+}
+
+impl DBTransaction {
+    pub fn put<K: AsRef<[u8]>, V: AsRef<[u8]>>(&mut self, col: usize, key: K, value: V) {
+        self.ops.push(DBOp::Insert {
+            col,
+            key: key.as_ref().to_owned(),
+            value: value.as_ref().to_owned(),
+        });
+    }
+
+    pub fn delete<K: AsRef<[u8]>>(&mut self, col: usize, key: K) {
+        self.ops.push(DBOp::Delete { col, key: key.as_ref().to_owned() });
+    }
+}
+
+pub struct RocksDB {
+    db: RwLock<Option<DB>>,
```

Comment:
Make sense

Reply:
Yeah, I confirm from rocksdb doc:
> A database may only be opened by one process at a time. The rocksdb implementation acquires a lock from the operating system to prevent misuse. Within a single process, the same rocksdb::DB object may be safely shared by multiple concurrent threads. I.e., different threads may write into or fetch iterators or call Get on the same database without any external synchronization (the rocksdb implementation will automatically do the required synchronization)